app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionActivity.kt

```diff
@@ -0,0 +1,106 @@
+    lateinit var presenter: ProgramStageSelectionPresenter
+
+    lateinit var binding: ActivityProgramStageSelectionBinding
+
+    lateinit var adapter: ProgramStageSelectionAdapter
+    private lateinit var enrollmentId: String
+    private lateinit var programId: String
+
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        programId = intent.getStringExtra(PROGRAM_UID)
+        enrollmentId = intent.getStringExtra(ENROLLMENT_UID)
+        val eventCreationType = intent.getStringExtra(EVENT_CREATION_TYPE)
+        (applicationContext as App).userComponent()?.plus(
+            ProgramStageSelectionModule(
+                this,
+                programId,
+                enrollmentId,
+                eventCreationType
+            )
+        )?.inject(this)
+        super.onCreate(savedInstanceState)
+
+        binding = DataBindingUtil.setContentView(this, R.layout.activity_program_stage_selection)
+        binding.presenter = presenter
+        adapter = ProgramStageSelectionAdapter(presenter)
+        binding.recyclerView.adapter = adapter
+    }
+
+    override fun onResume() {
+        super.onResume()
+        val orientation = Resources.getSystem().configuration.orientation
+        presenter.getProgramStages(programId, enrollmentId) // TODO: enrollment / event path
+        val columnCount = if (orientation == Configuration.ORIENTATION_LANDSCAPE) 3 else 2
+        binding.recyclerView.layoutManager = GridLayoutManager(this, columnCount)
+    }
+
+    override fun onPause() {
+        presenter.onDetach()
+        super.onPause()
+    }
+
+    override fun setData(programStages: List<ProgramStage>) {
+        if (programStages.isNotEmpty()) {
+            adapter.setProgramStages(programStages)
+            adapter.notifyDataSetChanged()
+        } else {
+            // if there are no program stages to select, the event cannot be added
+            displayMessage(getString(R.string.program_not_allow_events))
+            finish()
+        }
+    }
+
+    override fun setResult(programStageUid: String, repeatable: Boolean?, periodType: PeriodType?) {
+        val intent = Intent(this, EventInitialActivity::class.java)
+        val bundle = Bundle()
+        bundle.putString(PROGRAM_UID, getIntent().getStringExtra(PROGRAM_UID))
+        bundle.putString(
+            TRACKED_ENTITY_INSTANCE,
+            getIntent().getStringExtra(TRACKED_ENTITY_INSTANCE)
+        )
+        bundle.putString(ORG_UNIT, getIntent().getStringExtra(ORG_UNIT))
+        bundle.putString(ENROLLMENT_UID, getIntent().getStringExtra(ENROLLMENT_UID))
+        bundle.putString(EVENT_CREATION_TYPE, getIntent().getStringExtra(EVENT_CREATION_TYPE))
+        bundle.putBoolean(EVENT_REPEATABLE, repeatable ?: false)
+        bundle.putSerializable(EVENT_PERIOD_TYPE, periodType)
+        bundle.putString(Constants.PROGRAM_STAGE_UID, programStageUid)
+        bundle.putInt(EVENT_SCHEDULE_INTERVAL, presenter.getStandardInterval(programStageUid))
```

Comment:
You could use 'with' method or 'run' extesion function to avoid use all time 'bundle'

Reply:
Intent(this, EventInitialActivity::class.java).apply {
 intent.addFlags(Intent.FLAG_ACTIVITY_FORWARD_RESULT)
 //create bundle and fill it
 intent.putExtras(bundle)
}. also {
startActivity(it)
finish()
}

Reply:
this is one way... you have many different. If u do not want complications used marina's approach